pkg/ruler/rule_concurrency.go

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ruler
+
+import (
+	"context"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/tenant"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/rules"
+	"golang.org/x/sync/semaphore"
+)
+
+type MultiTenantRuleConcurrencyController interface {
```

Comment:
[nit] Why do we need this interface if it's just a `rules.RuleConcurrencyController`?

Reply:
The main reason is:

> There's another thing I forgot to mention. We never purge the tenantConcurrency in case of tenant inactivity. I don't think we have to address it in this PR, but I suggest it to address it in a following PR. I have a relatively simple idea on how to do it, we can talk on Slack about it.

I'm going to need at least one more method to purge semaphores of tenants the rulers we no longer own, and to do so I _think_ I'm gonna have to do it in `ruler.go`. Hence the interface I need and the prometheus one will no longer match. 

I can remove it if you want to given I'm OK with addressing the purging in a separate PR (my plan so to eventually add it to this one).

Reply:
It's not a big deal. We can keep the interface.